### PR TITLE
[JENKINS-29844] Make PMD Plugin compatible with Workflow

### PR DESCRIPTION
**Pull request by recena**
Downstream of https://github.com/jenkinsci/analysis-core-plugin/pull/31

Make PMD Plugin compatible with Workflow [JENKINS-29844](https://issues.jenkins-ci.org/browse/JENKINS-29844)

```
node {
  checkout changelog: false, poll: false, scm: [$class: 'SubversionSCM', additionalCredentials: [], excludedCommitMessages: '', excludedRegions: '', excludedRevprop: '', excludedUsers: '', filterChangelog: false, ignoreDirPropChanges: false, includedRegions: '', locations: [[credentialsId: 'efaa090d-5627-4a5c-95df-f570850d89a1', depthOption: 'infinity', ignoreExternalsOption: true, local: '.', remote: 'https://server/svn/repo/trunk']], workspaceUpdater: [$class: 'UpdateUpdater']]
  sh 'mvn clean install -Dmaven.test.skip pmd:check'
  step([$class: 'PmdPublisher'])
}
```

- [x] Initial source code modifications
- [x] Add an automated test specific for Workflow compatibility

### CODE REVIEW

**Comment by recena** [comment]
@reviewbybees Could you review these changes? Specially, @uhafner

File: src/main/java/hudson/plugins/pmd/PmdReporterResult.java [hunk 1]
```diff
@@ -35,6 +36,27 @@ public PmdReporterResult(final AbstractBuild<?, ?> build, final String defaultEn
...
+     * @param result
+     *            the parsed result with all annotations
+     * @param usePreviousBuildAsReference
+     *            determines whether to use the previous build as the reference
+     *            build
+     * @param useStableBuildAsReference
+     *            determines whether only stable builds should be used as
+     *            reference builds or not
+     */
+    public PmdReporterResult(final Run<?, ?> build, final String defaultEncoding, final ParserResult result,
```
Comment (amuniz) [inline src/main/java/hudson/plugins/pmd/PmdReporterResult.java]: The original constructor must be ```@Deprecated``` and call this one directly.

File: src/main/java/hudson/plugins/pmd/PmdResult.java [hunk 2]
```diff
@@ -41,14 +45,38 @@ public PmdResult(final AbstractBuild<?, ?> build, final String defaultEncoding,
...
+     * @param result
+     *            the parsed result with all annotations
+     * @param usePreviousBuildAsReference
+     *            determines whether to use the previous build as the reference
+     *            build
+     * @param useStableBuildAsReference
+     *            determines whether only stable builds should be used as
+     *            reference builds or not
+     */
+    public PmdResult(final Run<?, ?> build, final String defaultEncoding, final ParserResult result,
```
Comment (amuniz) [inline src/main/java/hudson/plugins/pmd/PmdResult.java]: Call this constructor from the original

File: src/main/java/hudson/plugins/pmd/PmdResult.java [hunk 3]
```diff
@@ -57,6 +85,28 @@ protected PmdResult(final AbstractBuild<?, ?> build,
...
+     * Creates a new instance of {@link PmdResult}.
+     *
+     * @param build the current build as owner of this action
+     * @param defaultEncoding the default encoding to be used when reading and parsing files
+     * @param result the parsed result with all annotations
+     * @param usePreviousBuildAsReference the value of usePreviousBuildAsReference
+     * @param useStableBuildAsReference determines whether only stable builds should be used as reference builds or not
+     * @param actionType the type of the result action
+     */
+    protected PmdResult(final Run<?, ?> build,
```
Comment (amuniz) [inline src/main/java/hudson/plugins/pmd/PmdResult.java]: Call this constructor from the original

File: src/main/java/hudson/plugins/pmd/PmdResult.java [hunk 4]
```diff
@@ -66,6 +116,15 @@ protected PmdResult(final AbstractBuild<?, ?> build,
         }
     }
 
+    PmdResult(final Run<?, ?> build, final BuildHistory history,
```
Comment (amuniz) [inline src/main/java/hudson/plugins/pmd/PmdResult.java]: Call this constructor from the original

File: src/main/java/hudson/plugins/pmd/PmdResultAction.java [hunk 5]
```diff
@@ -27,10 +29,24 @@
...
+     * Creates a new instance of <code>PmdResultAction</code>.
+     *
+     * @param owner
+     *            the associated build of this action
+     * @param healthDescriptor
+     *            health descriptor to use
+     * @param result
+     *            the result in this build
+     */
+    public PmdResultAction(final Run<?, ?> owner, final HealthDescriptor healthDescriptor, final PmdResult result) {
```
Comment (amuniz) [inline src/main/java/hudson/plugins/pmd/PmdResultAction.java]: Call this constructor from the original

**Comment by amuniz** [comment]
Missing tests for ```FAILURE``` and ```UNSTABLE``` status.

File: src/main/java/hudson/plugins/pmd/PmdPublisher.java [hunk 6]
```diff
@@ -129,17 +143,25 @@ public String getPattern() {
...
+    }
+
     @Override
     public Action getProjectAction(final AbstractProject<?, ?> project) {
         return new PmdProjectAction(project);
     }
 
     @Override
-    public BuildResult perform(final AbstractBuild<?, ?> build, final PluginLogger logger) throws InterruptedException, IOException {
+    public BuildResult perform(final Run<?, ?> build, final FilePath workspace, final PluginLogger logger) throws InterruptedException, IOException {
```
Comment (tfennelly) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: Again, I think you should add a `@Deprecated` override for these methods. It's hard to imagine someone extending these or calling directly, but you never know.

Comment (amuniz) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: @tfennelly answered here https://github.com/jenkinsci/checkstyle-plugin/pull/8/files#r36499970

Comment (tfennelly) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: @amuniz thanks Antonio.

**Comment by tfennelly** [comment]
:bee: from me

I agree with @amuniz on the constructor calls, but your call imo. Same for the tests, although maybe the test you have is enough to cover the main flow.

**Comment by recena** [comment]
@tfennelly Thanks for your feedback. By now, before to add more tests, I prefer to wait for @uhafner 's reviews.

**Comment by amuniz** [comment]
I see no bugs, but I'm not going to give a bee until additional tests are added :smile_cat:

**Comment by recena** [comment]
@amuniz the WAR has started!

File: src/main/java/hudson/plugins/pmd/PmdPublisher.java [hunk 7]
```diff
@@ -25,14 +24,20 @@
  * @author Ulli Hafner
  */
 public class PmdPublisher extends HealthAwarePublisher {
+
     private static final long serialVersionUID = 6711252664481150129L;
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: @stephenc should it be changed?

File: src/main/java/hudson/plugins/pmd/PmdResultAction.java [hunk 8]
```diff
@@ -27,10 +29,24 @@
      * @param result
      *            the result in this build
      */
+    @Deprecated
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/pmd/PmdResultAction.java]: deprecated javadoc?

File: src/main/java/hudson/plugins/pmd/PmdPublisher.java (hunk 7 above)
Comment (stephenc) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: No never ever change a `serialVersionUID` unless you cannot deserialize a correct object from the older format.

An example of when to change:
* You add a `NonNull` field and do not provide a `readResolve` method that enforces the non-null when old versions are deserialize and you want to screw your users over in strange and obtuse ways.

In Jenkins you basically *never* want to change the value, hence why you should just start every `Serializable` with the value `1L` so that we know from the start what the value is. The only time to use a value that is not `1L` is when you shipped by mistake... then you use the value that was shipped and leave it alone thereafter.

Comment (KostyaSha) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: How verify deserialisation?

Comment (stephenc) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: Capture a sample instance in a `byte[]` from before the change and then use that as test data, verifying that the `readResolve()` does the correct transformation / populating of fields... but you'd need a test like that whenever you add a `readResolve()` anyway

Comment (stephenc) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: To clarify, you need to capture the `byte[]` and then copy and paste the bytes into your test source / resources

Comment (KostyaSha) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: Do you mean only XStream deserialization?

Comment (stephenc) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: Nope. XStream doesn't care about serialVersionUID. Remoting uses java serialization and thus if you end up with a different JVM on the other side that may calculate a different default value and cause a specific slave to fail to be able to capture PMD results for example

File: src/main/java/hudson/plugins/pmd/PmdResultAction.java (hunk 8 above)
Comment (recena) [inline src/main/java/hudson/plugins/pmd/PmdResultAction.java]: @KostyaSha done

File: src/main/java/hudson/plugins/pmd/PmdPublisher.java (hunk 7 above)
Comment (KostyaSha) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: Hm, can't find general info about readResolve() for serialisation. Is it remoting specific?

Comment (KostyaSha) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: http://docs.oracle.com/javase/8/docs/platform/serialization/spec/input.html#a5903 this?

Comment (stephenc) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: That's the generic form.

Basically any time you change a class that gets serialized **and** your class introduces new invariants
* e.g. a field which cannot be `null`... there are other potential invariants, but in Jenkins they will all have to be implemented as a field that should not be `null` but is `null` when an older serialized form is deserialized in a newer class.
Well you then need to implement a `readResolve()` method to restore the invariants on deserialization. That `readResolve()` will be called by both XStream and regular java serialization. (you can get more fancy and use other forms of serialization control but XStream can have limited support for some of them).

The most typical use case is where you decide that you want to change the data format... in which case - to handle the transition correctly - you would make the old format fields `transient`. In the `readResolve()` method, if the old form was serialized then the values from the `transient` fields will be populated and thus you check if they were non-`null` and do the translation and populate the new fields (which in this case will be `null` as it was the old serialized form)

The second case is where you want to add a new field that should not be null (now we could argue that a better way is just to have the getter for that field initialize it if it was `null` and always access via the getter... but synchronization costs may mean you might prefer to have the field `/* effectively final */` and thus prefer using the `readResolve()` where you can ignore synchronization as the object instance has not escaped the "constructor" that is deserialization and thus cannot be seen by any other threads - screw you findbugs with your inconsistent synchronization warnings)

HTH

Comment (stephenc) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: If you look carefully in some of the well written long evolved plugins you will see the "fun" "dancing games" that happen in `readResolve()` methods and you will also see some `/*effectively final*/` comments where the field would be regarded as `final` but because it needs to be populated from the `readResolve()` method it isn't... an alternative method would be to have the `readResolve()` return a replacement instance... but beware class evolution as it is very easy to forget to update the copy constructor used in the `readResolve()` when you add a new field... so sometimes it is safer to have the field effectively final (i.e. only written from either constructors or `readResolve()`)

Comment (KostyaSha) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: For xml I will prefer to have automatic default value field creation if it exist, but we have what we have...

**Comment by tfennelly** [comment]
:bee:

**Comment by amuniz** [comment]
:bee:

**Comment by recena** [comment]
@reviewbybees done

**Comment by jglick** [comment]
a belated :bee:

**Comment by recena** [comment]
@jglick your :bee: are never belated :)

**Comment by uhafner** [comment]
See findbugs comments about deprecated methods and assertTrue vs. assertEquals.

File: pom.xml [hunk 9]
```diff
@@ -4,17 +4,19 @@
...
   <artifactId>pmd</artifactId>
   <packaging>hpi</packaging>
   <name>PMD Plug-in</name>
   <version>3.42-SNAPSHOT</version>
-  <url>http://wiki.jenkins-ci.org/x/GAAHAQ</url>
-  <description>This plug-in generates the trend report for PMD, an
-    open source static code analysis program.
-  </description>
+  <url>https://wiki.jenkins-ci.org/display/JENKINS/PMD+Plugin</url>
+  <description>This plug-in generates the trend report for PMD, an open source static code analysis program.</description>
```
Comment (uhafner) [inline pom.xml]: no reformatting.

File: src/main/java/hudson/plugins/pmd/PmdPublisher.java [hunk 10]
```diff
@@ -25,14 +24,20 @@
  * @author Ulli Hafner
  */
 public class PmdPublisher extends HealthAwarePublisher {
+
```
Comment (uhafner) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: remove

File: pom.xml (hunk 9 above)
Comment (KostyaSha) [inline pom.xml]: What initial URL is about?

File: src/main/java/hudson/plugins/pmd/PmdPublisher.java [hunk 11]
```diff
@@ -25,14 +24,20 @@
...
     private static final long serialVersionUID = 6711252664481150129L;
 
     private static final String PLUGIN_NAME = "PMD";
 
-    /** Default PMD pattern. */
+    /**
+     * Default PMD pattern.
+     */
```
Comment (uhafner) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: no reformat.

File: pom.xml (hunk 9 above)
Comment (uhafner) [inline pom.xml]: It's the tiny URL that will never change: now if someone renames the URL it will break :-(

File: src/main/java/hudson/plugins/pmd/PmdPublisher.java (hunk 11 above)
Comment (KostyaSha) [inline src/main/java/hudson/plugins/pmd/PmdPublisher.java]: Initial comment is wrong for javadoc

File: pom.xml (hunk 9 above)
Comment (uhafner) [inline pom.xml]: See also: https://confluence.atlassian.com/display/DOC/Configuring+Shortcut+Links

File: pom.xml [hunk 12]
```diff
@@ -37,7 +37,7 @@
     <dependency>
       <groupId>org.jvnet.hudson.plugins</groupId>
       <artifactId>analysis-core</artifactId>
-      <version>1.73-SNAPSHOT</version>
+      <version>1.73-beta-SNAPSHOT</version>
```
Comment (KostyaSha) [inline pom.xml]: shouldn't be changed?

File: pom.xml (hunk 9 above)
Comment (KostyaSha) [inline pom.xml]: AFAIR confluence preserve initial page name

Comment (KostyaSha) [inline pom.xml]: Found, seems permalink is the right way.
(imho  even if url changed you can change it for next releases).

Comment (KostyaSha) [inline pom.xml]: If description is no longer then 120 then imho change for description makes sense (but up to you of course)

**Comment by recena** [comment]
@uhafner all your comments have been addressed and the PR squashed.

File: .gitignore [hunk 13]
```diff
@@ -1 +1,2 @@
 target
+.idea/
```
Comment (KostyaSha) [inline .gitignore]: imho Such changes better do in separate commit, they are background.

Comment (recena) [inline .gitignore]: @KostyaSha I agree. I'll wait to @uhafner comments.

**Comment by recena** [comment]
@reviewbybees done
